connectors/sync_job_runner.py

```diff
@@ -0,0 +1,175 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+import asyncio
+import time
+
+from connectors.byoc import JobStatus
+from connectors.es import Mappings
+from connectors.logger import logger
+
+
+class SyncJobRunningError(Exception):
+    pass
+
+
+class JobClaimError(Exception):
+    pass
+
+
+class SyncJobRunner:
+    """
+    The class to run a sync job.
+
+    It takes a sync job, and tries to claim it, and executes it. It also makes sure the sync job is updated
+    appropriately when it errors out, is canceled, completes successfully or the service shuts down.
+
+    :param source_klass: The source class of the connector
+    :param sync_job: The sync job to run
+    :param connector: The connector of the sync job
+    :param elastic_server: The sync orchestrator used to fetch data from 3rd-party source and ingest into Elasticsearch
+    :param bulk_options: The bulk options used for the ingestion
+
+    """
+
+    def __init__(
+        self,
+        source_klass,
+        sync_job,
+        connector,
+        elastic_server,
+        bulk_options,
+    ):
+        self.source_klass = source_klass
+        self.data_provider = None
+        self.sync_job = sync_job
+        self.job_id = self.sync_job.id
+        self.connector = connector
+        self.connector_id = self.connector.id
+        self.elastic_server = elastic_server
+        self.bulk_options = bulk_options
+        self._start_time = None
+        self.running = False
+
+    async def execute(self):
+        if self.running:
+            raise SyncJobRunningError(f"Sync job #{self.job_id} is already running.")
+
+        self.running = True
+        if not await self._claim_job():
+            logger.error(f"Unable to claim job #{self.job_id}")
+            raise JobClaimError
+
+        try:
+            self.data_provider = self.source_klass(self.sync_job.configuration)
+            if not await self.data_provider.changed():
+                logger.debug(
+                    f"No change in {self.sync_job.service_type} data provider, skipping..."
+                )
+                await self._sync_done(sync_status=JobStatus.COMPLETED, result={})
+                return
+
+            logger.debug(f"Validating configuration for {self.data_provider}")
```

Comment:
I would also add the `connector_id` to that log message for better traceability.

Reply:
For this line I copied from here: https://github.com/elastic/connectors-python/pull/552#discussion_r1117323803

I think we can leave it.